python/ray/serve/api.py

```diff
@@ -1510,155 +1516,394 @@ def get_deployment_statuses() -> Dict[str, DeploymentStatusInfo]:
-                the Serve REST API schema.
 
-        Returns:
-            Application: a new application object containing the deployments.
-        """
-        raise NotImplementedError()
+class UnboundDeploymentMethodNode(DAGNode):
+    def __init__(
+        self,
+        deployment: Deployment,
+        method_name: str,
+        other_args_to_resolve: Dict[str, Any] = None,
+    ):
+        self._deployment = deployment
+        self._method_name = method_name
+        self._options = {}
+        self._bound_other_args_to_resolve = other_args_to_resolve
+
+    def bind(self, *args, **kwargs):
+        node = DeploymentMethodNode(
+            self._deployment,
+            self._method_name,
+            args,
+            kwargs,
+            self._options,
+            other_args_to_resolve=self._bound_other_args_to_resolve,
+        )
+        return node
 
-    def to_yaml(self, f: Optional[TextIO] = None) -> Optional[str]:
-        """Returns this application's deployments as a YAML string.
+    def options(self, **options):
+        self._options = options
+        return self
 
-        Optionally writes the YAML string to a file as well. To write to a
-        file, use this pattern:
 
-        with open("file_name.txt", "w") as f:
-            app.to_yaml(f=f)
+@PublicAPI(stability="alpha")
+class DeploymentMethodNode(DAGNode):
+    """Represents a method call on a bound deployment node.
 
-        This file is formatted as a Serve YAML config file. It can be deployed
-        via the Serve CLI.
+    These method calls can be composed into an optimized call DAG and passed
+    to a "driver" deployment that will orchestrate the calls at runtime.
 
-        Args:
-            f (Optional[TextIO]): A pointer to the file where the YAML should
-                be written.
+    This class cannot be called directly. Instead, when it is bound to a
+    deployment node, it will be resolved to a DeployedCallGraph at runtime.
+    """
 
-        Returns:
-            Optional[String]: The deployments' YAML string. The output is from
-                yaml.safe_dump(). Returned only if no file pointer is passed in.
-        """
-        raise NotImplementedError()
+    def __init__(
+        self,
+        deployment: Deployment,
+        deployment_method_name: str,
+        method_args: Tuple[Any],
+        method_kwargs: Dict[str, Any],
+        method_options: Dict[str, Any],
+        other_args_to_resolve: Optional[Dict[str, Any]] = None,
+    ):
+        self._deployment = deployment
+        self._deployment_method_name: str = deployment_method_name
+        super().__init__(
+            method_args,
+            method_kwargs,
+            method_options,
+            other_args_to_resolve=other_args_to_resolve,
+        )
+        self._deployment_handle: Union[
+            RayServeHandle, RayServeSyncHandle
+        ] = self._get_serve_deployment_handle(deployment, other_args_to_resolve)
```

Comment:
We shouldn't be actually getting a `ServeHandle` until later when we `run` or this is passed into a deployment. This causes serve to be initialized.